### ISSUE TITLE
Additional CASCI Methods

### ISSUE THREAD

**Issue by matillda123** [pyscf/pyscf#2340]
Hi,

I am doing potential energy surface scans of excited states. And cannot find methods to improve the results other than NEVPT2.
It would be nice if one could use casci methods beyond "standard" casci. (e.g. IVO-CASCI or CISNO-CASCI or something like that)

Thanks. :)

**Comment by JangidBhavnesh** [comment]
I think one can generate IVOs, or CISNOs before hand for the given excited state and then these orbitals can be used for the CASCI calculations.

Something like this ..

CISNO = .... # Generated from the TDA or CIS method of PySCF

mc = mcscf.CASCI(...)
mc.kernel(CISNO)

**Comment by MatthewRHermes** [comment]
Yeah the optional first argument of the CASCI kernel function is a set of MOs, to which you can pass any array you please. If I understand IVO's correctly:

```
mo_occ_ivo = mf.mo_occ.copy () # Important to copy, not edit in-place
nocc = numpy.count_nonzero (mo_occ_ivo>0) # At any rate, figure out, i.e., where the HOMO is
mo_occ_ivo[nocc-1] -= 1 # subtract an electron from HOMO
dm_ivo = mf.make_rdm1 (mf.mo_coeff, mo_occ_ivo) # (N-1)-electron density matrix
fock_ivo = mf.get_fock (dm=dm_ivo) # (N-1)-electron Fock operator
mo_ivo = mf.canonicalize (mf.mo_coeff, mf.mo_occ, fock=fock_ivo)[1] # Diagonalize Fock w/o mixing occ/vir
mo_ivo[:,:nocc] = mf.mo_coeff[:,:nocc] # Only keep the new virtual orbitals

mc = mcscf.CASCI (mf, ...)
mc.kernel (mo_ivo) # IVO-CASCI (if I understand correctly)
```

**Comment by matillda123** [comment]
Thank you Matthew. This seems to work as it should. :D

> Yeah the optional first argument of the CASCI kernel function is a set of MOs, to which you can pass any array you please. If I understand IVO's correctly:
> 
> ```
> mo_occ_ivo = mf.mo_occ.copy () # Important to copy, not edit in-place
> nocc = numpy.count_nonzero (mo_occ_ivo>0) # At any rate, figure out, i.e., where the HOMO is
> mo_occ_ivo[nocc-1] -= 1 # subtract an electron from HOMO
> dm_ivo = mf.make_rdm1 (mf.mo_coeff, mo_occ_ivo) # (N-1)-electron density matrix
> fock_ivo = mf.get_fock (dm=dm_ivo) # (N-1)-electron Fock operator
> mo_ivo = mf.canonicalize (mf.mo_coeff, mf.mo_occ, fock=fock_ivo)[1] # Diagonalize Fock w/o mixing occ/vir
> mo_ivo[:,:nocc] = mf.mo_coeff[:,:nocc] # Only keep the new virtual orbitals
> 
> mc = mcscf.CASCI (mf, ...)
> mc.kernel (mo_ivo) # IVO-CASCI (if I understand correctly)
> ```

**Comment by matillda123** [comment]
I tried implementing CISNO-CASCI based on the density matrices generated in example [https://github.com/pyscf/pyscf/blob/master/examples/tddft/22-density.py](url). But the resulting energies are very wrong. 

As far as I understand CISNO one generates natural orbitals from the state average density matrix from a TDDFT calculation. So what I did is:
```

dm=0
for i in range(nstates):
      dm=dm+get_density_matrix(td,i)  # using the function from example 22

dm=(mf.make_rdm1()+dm)/(nstates+1)
occ,no=scipy.linalg.eigh(dm)

# and then occ,no is reordered to get the occupied orbitals first. 

```

For now I'm quite satisfied with the performance of the IVOs. But If anyone knows whats wrong that would be nice.

**Comment by hebrewsnabla** [comment]
It's wrong to diagonalize AO-based DM directly because AOs are not orthogonal. One of the correct way of obtaining natural orbitals is 
https://github.com/pyscf/pyscf/blob/ca8c7c1680defdfee2380eda3af3a28d9fb375cb/pyscf/mcscf/addons.py#L350-L358

**Comment by matillda123** [comment]
Ah thank you @hebrewsnabla :D
It seems to work now.
